applications/FluidDynamicsApplication/custom_elements/two_fluid_navier_stokes.cpp

```diff
@@ -0,0 +1,2054 @@
+				Kee_tot.resize(NumNodes, NumNodes, false);
+				rhs_ee_tot.resize(NumNodes, false);
+
+                noalias(Vtot) = ZeroMatrix(NumNodes*(Dim + 1), NumNodes);
+                noalias(Htot) = ZeroMatrix(NumNodes, NumNodes*(Dim + 1));
+                noalias(Kee_tot) = ZeroMatrix(NumNodes, NumNodes);
+                noalias(rhs_ee_tot) = ZeroVector(NumNodes);
+
+                for (unsigned int g_pos = 0; g_pos < data.w_gauss_pos_side.size(); g_pos ++) {
+                    data.UpdateGeometryValues(
+                        g_pos,
+                        data.w_gauss_pos_side[g_pos],
+                        row(shape_functions_pos, g_pos),
+				        shape_derivatives_pos[g_pos],
+                        row(shape_functions_enr_pos,g_pos),
+                        shape_derivatives_enr_pos[g_pos]);
+                    
+                    data.CalculateDensityAtGaussPoint();
+                    data.CalculateAirMaterialResponse();
+                    this->AddTimeIntegratedSystem(
+				 		data, rLeftHandSideMatrix, rRightHandSideVector);
+                    
+                    ComputeGaussPointEnrichmentContributions(data, Vtot, Htot, Kee_tot, rhs_ee_tot);
+                }
+
+                for (unsigned int g_neg = 0; g_neg < data.w_gauss_neg_side.size(); g_neg ++) {
+                    data.UpdateGeometryValues(
+                        g_neg,
+                        data.w_gauss_neg_side[g_neg],
+                        row(shape_functions_neg, g_neg),
+				        shape_derivatives_neg[g_neg],
+                        row(shape_functions_enr_neg,g_neg),
+                        shape_derivatives_enr_neg[g_neg]);
+                    
+                    data.CalculateDensityAtGaussPoint();
+                    this->CalculateMaterialResponse(data);
+                    this->AddTimeIntegratedSystem(
+				 		data, rLeftHandSideMatrix, rRightHandSideVector);
+                    
+                    ComputeGaussPointEnrichmentContributions(data, Vtot, Htot, Kee_tot, rhs_ee_tot);
+                }
+
+				CondenseEnrichment(data, rLeftHandSideMatrix,rRightHandSideVector,Htot,Vtot,Kee_tot, rhs_ee_tot);
+			}
+		}
+
+		else {
+			//Get Shape function data 
+            Vector gauss_weights; 
+            Matrix shape_functions; 
+            ShapeFunctionDerivativesArrayType shape_derivatives; 
+            this->CalculateGeometryData( 
+              gauss_weights, shape_functions, shape_derivatives); 
+            const unsigned int number_of_gauss_points = gauss_weights.size(); 
+			// Iterate over integration points to evaluate local contribution
+			for (unsigned int g = 0; g < number_of_gauss_points; g++) {
+
+				data.UpdateGeometryValues(g, gauss_weights[g], row(shape_functions, g),
+					shape_derivatives[g]);
+
+                data.CalculateDensityAtGaussPoint();
+
+				if (data.IsAir())
+					data.CalculateAirMaterialResponse();
+				else
+					this->CalculateMaterialResponse(data);
```

Comment:
Again, maybe it would be cleaner to move the check inside this->CalculateMaterialResponse(), so that you don't need to modify the global interface because of your data's details.

Reply:
done